### PR TITLE
Milestone 9 : Recipe App | Final

### PR DESCRIPTION
**Pull request by michael-duke**
## Recipe App milestone 9 | Final Project

- Used Ruby gems as a software package system.
- Installed Ruby on Rails framework, and included Material Tailwind CDN for styling
- Understood Rails RESTful design and router.
- Used controllers to handle requests and render empty views.
- Understood Rails naming conventions.
- Used params from browser requests in a safe way.
- Used preprocessed HTML file with embedded Ruby code.
- Used layouts and templates for shared content.
- Used database migration files to maintain database schema.
- Used validations for models.
- Secured app from n+1 problems.
- Understood what ORM is.
- Written SQL queries with ActiveRecord.
- Set up associations between models.
- Built a web app that requires the user to log in.
- Described the difference between authorization and authentication.
- Used Devise gem for authentication.
- Understood how sessions and cookies can support authentication.
- Limited access to web app resources based on authorization rules.
- Implemented tests for models, controllers and views

## Future improvements
 - Notifications for the alerts, errors and notices
 - Make responsive UI that will tell the user if we runs out of ingredients or about to
## ⚠️ To Run our project 
- Make sure to create a `.env ` file that will be used for database migrations of development and test : 
```
POSTGRES_USER=YourUsername
# If you declared a password when creating the database:
POSTGRES_PASSWORD=YourPassword
POSTGRES_HOST=localhost
POSTGRES_DB=Recipe_App_development
POSTGRES_TEST_DB=Recipe_App_test
```
### Hapyy reviewing 😄

### CODE REVIEW

File: app/views/public_recipes/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<h1 class="text-2xl text-center mb-3">Public Recipes</h1>
+
+<div class="flex flex-col w-[40%] mx-auto">
+  <% @public_recipes.each do |recipe| %>
+    <%= render recipe %>
+  <% end %>
+</div>
```
Comment (omarsalem7) [inline app/views/public_recipes/index.html.erb]: - [x] Kindly make your public recipes match the wireframe design and contain the same information

| Your design  | Wireframe design|
| ------------- | ------------- |
|image  | image |

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    #
+    user ||= User.new # guest user (not logged in)
+
+    if user.present?
+      can(:manage, Food, user:)
+      can(:manage, RecipeFood, user:)
+      can :manage, Recipe, user:
+    else
+      can :read, :all # user(not admin) can read all the blogs
+    end
```
Comment (omarsalem7) [inline app/models/ability.rb]: - [x] When I register with the new user, I can't add food or recipe and I have the same authorization rules like the guest user, you can check [this video](https://www.loom.com/share/37b69ec06871425995c099cec5b7d33b) to get what I mean.

**Review by omarsalem7 — CHANGES_REQUESTED** [review]
Hi @michael-duke, @samiullah997 😊, 

Good job so far! 👏
You have done well reaching this point of the milestone, but there are some issues that you still need to work on to go to the next project but you are almost there! 

you-are-near

## Highlights
- Great design ✔️
- The Pull Request has a proper title and description. ✔️
- Correct git-flow ✔️

## Required Changes ♻️
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/public_recipes/index.html.erb (hunk 1 above)
Comment (michael-duke) [reply app/views/public_recipes/index.html.erb]: I will make the changes thank you @omarsalem7

File: app/models/ability.rb (hunk 2 above)
Comment (michael-duke) [reply app/models/ability.rb]: I will look into it, thank you for making the video 😄 @omarsalem7

Comment (omarsalem7) [reply app/models/ability.rb]: You are welcome @michael-duke 🌸

File: app/views/public_recipes/index.html.erb (hunk 1 above)
Comment (michael-duke) [reply app/views/public_recipes/index.html.erb]: Changes made and fixed

File: app/models/ability.rb (hunk 2 above)
Comment (michael-duke) [reply app/models/ability.rb]: Changes done, thank you for your feedback

**Review by omarsalem7 — APPROVED** [review]
Hi @michael-duke, @samiullah997 😊, 
# STATUS: APPROVED! 🥇 💯 

giphy

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 🚀 
Congratulations! 🎉 

## Highlights
✅ Good descriptive pull request 
✅ Project has met all the technical aspects you needed for this project. Keep rocking! 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by michael-duke** [comment]
Thank you @omarsalem7 for you feedback, and for helping us improve our project , happy coding 😄 🚀

**Pull request merged**